### ISSUE TITLE
Intellisense CPU usage high

### ISSUE THREAD

**Issue by jrieken** [OmniSharp/omnisharp-roslyn#367]
_From @h0rdak on December 16, 2015 8:54_

Whenever I use intellisense in my C# projects, the CPU usage spikes and makes my fans go crazy and my laptop unbearably hot to the point it's completely unusable.

This is with vs code 0.10.3 on OSX 10.10.5, on a macbook pro. 

![screenshot 2015-12-16 09 46 20](https://cloud.githubusercontent.com/assets/7004085/11836275/e1181bcc-a3d9-11e5-81f0-89ce53df1292.png)

Activity monitor points to vs code -> electron helper -> mono-sgen

![screenshot 2015-12-16 09 44 49](https://cloud.githubusercontent.com/assets/7004085/11836260/c131ab52-a3d9-11e5-973c-fba72f7b4423.png)

Installed mono is:

Mono JIT compiler version 4.2.1 (Stable 4.2.1.102/6dd2d0d Wed Dec  2 14:02:16 PST 2015)

_Copied from original issue: Microsoft/vscode#1368_

**Comment by jrieken** [comment]
@h0rdak Any details you can share with us? Project size and kind (dnx, msbuild) Could you attach the OmniSharp Log (Cmd+L L)?

**Comment by h0rdak** [comment]
This is with an empty Unity project, using the integration plugin. 

Looking at the log, I seem to be getting an exception whenever intellisense is triggered:

System.ArgumentException: Diagnostic must have span '[128..128)' 

Attached log, I get a new one of these every time intellisense is triggered.

[vscodelog.txt](https://github.com/OmniSharp/omnisharp-roslyn/files/63853/vscodelog.txt)

**Comment by jrieken** [comment]
Thanks. The errors are unhappy but shouldn't be the cause of the issue. Can you also hint to the version of mono you are using?

**Comment by h0rdak** [comment]
Mono JIT compiler version 4.2.1 (Stable 4.2.1.102/6dd2d0d Wed Dec 2 14:02:16 PST 2015)

Was using 4.2.0 initially, ran into this and upgraded, still the same issue.

**Comment by h0rdak** [comment]
I seem to have fixed the issue by installing dnvm and using it to download and install dnx-mono.1.0.0-rc1-update1

**Comment by shochet** [comment]
Same problem here with slow intellisense and high cpu. Just started after rebooting this morning.
mono-sgen and Electron Helper both run out of control for 20 seconds whenever I trigger intellisense, or go to definition, Find all references, etc.

VSCode 0.10.5
Unity 5.1.4
Mono 4.2.0
Macbook Pro OSX 10.11.2

Driving me crazy...

**Comment by troydai** [comment]
Hi @jrieken. It has been a few months since you file this issue. Since then on `dev` branch, we introduced new tool-chain and update the `Roslyn` to `1.1.1`. Can you try out the `dev` branch build to see if this is still a huge issue?

**Comment by fishg** [comment]
How to upgrade VSCode's roslyn?

**Comment by troydai** [comment]
@fishg you can follow the instructions in this readme: https://github.com/OmniSharp/omnisharp-roslyn/blob/dev/README.md.

**Comment by fishg** [comment]
@troydai
Looks like build.sh have some problem on OS X:

Installing dotnet from beta channel for version 1.0.0.001897
Execute install script
   source: https://raw.githubusercontent.com/dotnet/cli/43ac2b45f4173b8228b44c8a9693ac7774104cbb/scripts/obtain/install.sh
  version: 1.0.0.001897
  channel: beta
  install: /Users/fishg/data/opensource/omnisharp-roslyn/.dotnet
./build.sh: line 35: /Users/fishg/data/opensource/omnisharp-roslyn/.dotnet/cli/dotnet: No such file or directory
*** Building ***
./build.sh: line 44: /Users/fishg/data/opensource/omnisharp-roslyn/.dotnet/cli/dotnet: No such file or directory
./build.sh: line 45: /Users/fishg/data/opensource/omnisharp-roslyn/.dotnet/cli/dotnet: No such file or directory
./build.sh: line 46: /Users/fishg/data/opensource/omnisharp-roslyn/.build/PublishProject/PublishProject: No such file or directory

**Comment by fishg** [comment]
@troydai Sorry, it's my network problems.

**Comment by fishg** [comment]
@troydai 
When Publishing, no response.How to resolve?

NuGet Config files used:
    /Users/fishg/data/opensource/omnisharp-roslyn/NuGet.Config
    /Users/fishg/.nuget/NuGet/NuGet.Config

Feeds used:
    https://www.myget.org/F/omnisharp-roslyn-ci/api/v3/index.json
Publishing

**Comment by fishg** [comment]
Publishing blocked long time.No any result.

**Comment by troydai** [comment]
@fishg can you copy the full log? I work on OSX full time so I doubt it is platform specific. It could be some problem with you miss some tools for packaging.

**Comment by fishg** [comment]
@troydai 
 ./build.sh
*** Cleanup ***
*** Installing dotnet ***
Installing dotnet from beta channel for version 1.0.0.001897
Execute install script
   source: https://raw.githubusercontent.com/dotnet/cli/43ac2b45f4173b8228b44c8a9693ac7774104cbb/scripts/obtain/install.sh
  version: 1.0.0.001897
  channel: beta
  install: /Users/fishg/data/opensource/omnisharp-roslyn/.dotnet
dotnet_install: Preparing to install .NET Tools from 'beta' channel to '/Users/fishg/data/opensource/omnisharp-roslyn/.dotnet'
dotnet_install: Downloading dotnet-dev-osx-x64.1.0.0.001897.tar.gz from https://dotnetcli.blob.core.windows.net/dotnet/beta/Binaries/1.0.0.001897
######################################################################## 100.0%
dotnet_install: Extracting tarball
dotnet_install: Moving new CLI into install location and symlinking
dotnet_install: Cleaning dotnet-dev-osx-x64.1.0.0.001897.tar.gz
.NET Command Line Tools (1.0.0-beta-001897)

Product Information:
 Version:     1.0.0-beta-001897
 Commit Sha:  ca74a9b177

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.11
 OS Platform: Darwin
 Runtime Id:  osx.10.11-x64
*** Building ***
log  : Restoring packages for /Users/fishg/data/opensource/omnisharp-roslyn/tools/PublishProject/project.json...
info : Committing restore...
log  : tools/PublishProject/project.json
log  : Restore completed in 1734ms.

NuGet Config files used:
    /Users/fishg/data/opensource/omnisharp-roslyn/NuGet.Config
    /Users/fishg/.nuget/NuGet/NuGet.Config

Feeds used:
    https://www.myget.org/F/omnisharp-roslyn-ci/api/v3/index.json
Publishing

No error log found.

**Comment by fishg** [comment]
When I deleted nuget's cache, I build successed.
But a new problem is:
dnxcore50 have a excutable OmniSharp file,but not work in vscode with Unity3D project.
dnx451 have no excutable Unix file,so I add a new sh file:
```
#!/bin/bash
/usr/local/bin/mono /Users/fishg/data/opensource/omnisharp-roslyn/artifacts/publish/OmniSharp/osx.10.11-x64/dnx451/OmniSharp.exe "$@"
```
Now works.
![qq20160323-0 2x](https://cloud.githubusercontent.com/assets/1423545/13980265/957e48f8-f118-11e5-8501-c5e02e28b7d8.png)

**Comment by fishg** [comment]
Cpu usage is also too high when open project or switch to new class file.
![qq20160323-1 2x](https://cloud.githubusercontent.com/assets/1423545/13980461/b8b84a66-f119-11e5-984c-543abe4a0fef.png)

**Comment by troydai** [comment]
@fishg can you elaborate what doesn't work with Unity under dnxcore50 OmniSharp?

**Comment by fishg** [comment]
@troydai If use dnxcore50,process start normally, but no code any intellisense.

**Comment by troydai** [comment]
@fishg can you point me to a mini repo for missing intellisense case?

**Comment by speoplesz** [comment]
This is still an issue.

**Comment by Jodacola** [comment]
I've been struggling with instances of intensely high CPU usage at odd points while coding in VSCode for over a year, now.  I previously noticed some patterns related to working on classes implementing `IDisposable`, and again hit the issue today, to much frustration.  I've managed to recreate the issue, albeit without the high CPU usage, in a simple project.  Environment, steps, and findings below.

**Operating environment:**

- MacBook Pro 15 Mid 2015
- 16 GB RAM
- macOS Mojave 10.14.4
- VSCode 1.33.1
- Mono JIT compiler version 5.16.0.221 (2018-06/b63e5378e38 Mon Nov 19 18:08:09 EST 2018)

**Steps to reproduce**

1. `dotnet new console -n MonoSgenCall`
2. `cd MonoSgenCall`
3. `code .`
4. Add required resources via VSCode prompt
5. Add a new file to the project named `SomeDisposableClass.cs` with following content:

```c#
using System;

public class SomeDisposableClass : IDisposable
{
    public void Dispose()
    {
    }
}
```
6. Start Activity Monitor to monitor process on CPU tab.
7. Make any compilable change to `SomeDisposableClass` and save the file and immediately note mono-sgen64 jumps up in the Activity Monitor list.
![Screen Shot 2019-04-20 at 10 07 28 AM](https://user-images.githubusercontent.com/3253637/56459059-4c4c9200-6354-11e9-9086-9a1969013dd2.png)
8. Update `SomeDisposableClass` to not implement `IDisposable`:
```c#
using System;

public class SomeDisposableClass
{
    public void Dispose()
    {
    }
}
```
9. Make changes to class and save, as before
10. Note that mono-sgen64 no longer jumps to the top of the list

CPU usage jumps far, far more when running this exercise (steps 5-7) in a much larger project:
![Screen Shot 2019-04-20 at 10 11 58 AM](https://user-images.githubusercontent.com/3253637/56459122-deed3100-6354-11e9-8718-ca7871080cdb.png)

This, again, only happens for me when working on classes implementing `IDisposable`.  This has been repeatable for a very long time for me, but I didn't take the time to collect information and steps until today.  I've Googled quite a bit for IDisposable-related issues in this repo, IDisposable-related issues with Sgen in Mono, and with just VSCode, to no avail.  Hopefully this can help move progress on resolving this in the right direction.

Happy to provide any additional information or testing.

**Comment by atliegils** [comment]
Nice post, @Jodacola   
Did you manage to find more information to resolve this issue?

**Comment by chinaq** [comment]
> I've been struggling with instances of intensely high CPU usage at odd points while coding in VSCode for over a year, now. I previously noticed some patterns related to working on classes implementing `IDisposable`, and again hit the issue today, to much frustration. I've managed to recreate the issue, albeit without the high CPU usage, in a simple project. Environment, steps, and findings below.
> 
> **Operating environment:**
> 
> * MacBook Pro 15 Mid 2015
> * 16 GB RAM
> * macOS Mojave 10.14.4
> * VSCode 1.33.1
> * Mono JIT compiler version 5.16.0.221 (2018-06/b63e5378e38 Mon Nov 19 18:08:09 EST 2018)
> 
> **Steps to reproduce**
> 
> 1. `dotnet new console -n MonoSgenCall`
> 2. `cd MonoSgenCall`
> 3. `code .`
> 4. Add required resources via VSCode prompt
> 5. Add a new file to the project named `SomeDisposableClass.cs` with following content:
> 
> ```cs
> using System;
> 
> public class SomeDisposableClass : IDisposable
> {
>     public void Dispose()
>     {
>     }
> }
> ```
> 
> 1. Start Activity Monitor to monitor process on CPU tab.
> 2. Make any compilable change to `SomeDisposableClass` and save the file and immediately note mono-sgen64 jumps up in the Activity Monitor list.
>    ![Screen Shot 2019-04-20 at 10 07 28 AM](https://user-images.githubusercontent.com/3253637/56459059-4c4c9200-6354-11e9-9086-9a1969013dd2.png)
> 3. Update `SomeDisposableClass` to not implement `IDisposable`:
> 
> ```cs
> using System;
> 
> public class SomeDisposableClass
> {
>     public void Dispose()
>     {
>     }
> }
> ```
> 
> 1. Make changes to class and save, as before
> 2. Note that mono-sgen64 no longer jumps to the top of the list
> 
> CPU usage jumps far, far more when running this exercise (steps 5-7) in a much larger project:
> ![Screen Shot 2019-04-20 at 10 11 58 AM](https://user-images.githubusercontent.com/3253637/56459122-deed3100-6354-11e9-8718-ca7871080cdb.png)
> 
> This, again, only happens for me when working on classes implementing `IDisposable`. This has been repeatable for a very long time for me, but I didn't take the time to collect information and steps until today. I've Googled quite a bit for IDisposable-related issues in this repo, IDisposable-related issues with Sgen in Mono, and with just VSCode, to no avail. Hopefully this can help move progress on resolving this in the right direction.
> 
> Happy to provide any additional information or testing.



This is still an issue that 3 years passed.

**Comment by SirIntruder** [comment]
Unfortunately this seems to behave bad in both Visual Studio  and Visual Studio Code. I hoped it was just an omnisharp issue, but looks like it is somewhere in Roslyn repo.

There is a slight difference tough - taking a random sample in omnisharp repo (MonoScriptGenerationProcess.Disposable()), omnisharp/vscode displays 32 references, and Visual Studio displays 23 references. 

Both VS and VSCode crank up to the ~100% cpu usages when this gets recalculated, with a caveat that VS seems smarter about not recalculating this anytime something gets changed in that file.

I have no idea if listing every call to Disposable() everywhere is intentional or not (32 references?), seems pretty useless IMHO...

**Comment by kruncher** [comment]
We found that tweaking the C# extension for VSCode to ignore `Dispose` and `GetEnumerator` helped a lot. The issue still occurs though, seems to be more problematic more recently with no apparent reason.

**Comment by galvesribeiro** [comment]
I have the same situation when building xUnit test projects when I have to add a fixture class that usually implements `IDisposable` :

![image](https://user-images.githubusercontent.com/4714040/68043369-e3a38200-fcb3-11e9-8979-3b7f1a4f7c85.png)

At: 

![image](https://user-images.githubusercontent.com/4714040/68043630-7a703e80-fcb4-11e9-946b-e150e3e965af.png)

Using: 

![image](https://user-images.githubusercontent.com/4714040/68043696-a25fa200-fcb4-11e9-98fc-d90b0bf14f05.png)

![image](https://user-images.githubusercontent.com/4714040/68043734-b73c3580-fcb4-11e9-808d-f8edc8ed6904.png)

So... It is almost a 4yrs old issue... Is there a plan to attack this bug?

**Comment by Ursanon** [comment]
Try setting:
`"csharp.referencesCodeLens.enabled": false`    
![Screenshot 2019-12-18 at 13 35 56](https://user-images.githubusercontent.com/8686134/71086964-5ea8f500-219b-11ea-83fd-0a98c604dfe6.png)

It will disable reference counting for `IDisposable` interface and hide `x references` info, but un-lag your computer.

**Comment by yakeer** [comment]
Also experiencing this bug, MacOS Catalina - Latest VSCode.
I disabled the CodeLens references - it makes it a little bit better - but the CodeLens feature is really useful!
Hope you'll finally fix it.

Thanks.

**Comment by reckhou** [comment]
Boom, same issue here! But I'm using Windows version of VSCode & OmniSharp, the latest version.

Tried to turn off CodeLens and it fixes the issue. But hopefully that could be fixed somehow.

**Comment by galvesribeiro** [comment]
Yeah... sad to see we're entering the 5th year without a solution for that... C# is the major reason why people goes VSCode but it seem like there is not much attention to it... Sorry but disable CodeLenses is not a viable workaround... It works for other non-mainstream languages so there is no reason why it shouldn't work for the main one...

**Comment by mvalero** [comment]
Am observing exactly the same issue since several years while using omnisharp in VSCode, in Windows 10 and with several .net Core 2.1 Projects.

I already mentioned this issue in https://github.com/OmniSharp/omnisharp-vscode/issues/3079 as I thought it could be related to the omnisharp implementation in vscode, but it seems to be rather a general omnisharp roslyn problem.

As stated in the omnisharp vscode ticket the problem is really easily reproducible when editing a class that implements `IDisposable` with the CPU going directly to 100% utilisation. At the moment `IDisposable` is commented out or removed from the class the CPU usage goes quickly back to normal.

**Comment by MichielHegemans** [comment]
A workaround for the problem is disabling CodeLens for CSharp. It may not be a good workaround in case you depend a lot on CodeLens, but for some reason it stops melting my computer and mono-sgen64 no longer uses absurd amounts of CPU (even within `IDisposable` classes)

![image](https://user-images.githubusercontent.com/27996963/81307252-c3001e80-9080-11ea-902d-e3d3fde98203.png)

**Comment by 01010101lzy** [comment]
Same issue here, Omnisharp eating up 7% CPU when idle after opening a ASP.NET Core project. I'm working on 3 different-sized projects here, and Omnisharp is consuming the same CPU share when idle regardless of project size.

![图片](https://user-images.githubusercontent.com/10259119/92690704-08589380-f374-11ea-8fe8-18e2eab2c704.png)

**Comment by filipw** [comment]
I will close this because it's become mostly obsolete and is now a mix of different things from different platform - but mostly this issue has been related to embedded Mono. The embedded Mono approach has been retired and removed from OmniSharp. OmniSharp only exists as .NET 6.0 build and global Mono build at this point.

If there are CPU issues in those builds, please open new issues.
